management_interface/management_interface/admin.py

```diff
@@ -49,6 +58,108 @@ def save_model(self, request, obj, form, change):
 class CareProviderLocationAdmin(admin.ModelAdmin):
     form = CareProviderLocationForm
 
+    def _is_csv_in_request_files(self, request) -> bool:
+        if len(request.FILES) == 0 or "csvfile" not in request.FILES:
+            return False
+        return True
+
+    def _is_csv_line_count_valid(self, csv_data_list) -> bool:
+        if len(csv_data_list) > SETTINGS.CSV_IMPORT_MAX_LINES:
+            return False
+        return True
+
+    def _is_csv_column_set_valid(self, csv_data_list) -> bool:
+        required_columns = ["NHS_NUMBER", "DOB", "FAMILY_NAME", "GIVEN_NAME", "PROVIDER_REFERENCE"]
+        for column in required_columns:
+            if column not in csv_data_list[0]:
+                return False
+        return True
+
+    def _create_bulk_care_recipients(self, csv_data_list, careproviderlocation_id, user_id) -> int:
+        care_recipient_created_count = 0
+        care_provider_location = CareProviderLocation.objects.get(pk=careproviderlocation_id)
+        for care_recipient_dict in csv_data_list:
+            if not CareRecipient.objects.filter(
+                provider_reference_id=care_recipient_dict["PROVIDER_REFERENCE"]
+            ).exists():
+                care_recipient = CareRecipient(
+                    care_provider_location=care_provider_location,
+                    nhs_number=care_recipient_dict["NHS_NUMBER"],
+                    provider_reference_id=care_recipient_dict["PROVIDER_REFERENCE"],
+                    created_by_id=user_id,
+                )
+                care_recipient.clean()
+                care_recipient.save()
+                if care_recipient.pk:
```

Comment:
You can re-import file as many times as you want. We are checking if the specific `provider_reference_id` exists and ignore it. We can add some logging too. Probably need @henrywrightPSC to add his 2p.

Reply:
It's going to be pretty common that we get 1/2 errors in each set (either due to PDS validation failure (v likely), or some issue in the CSV) and in a large file we actually don't want to go through the trouble of re-importing them all cause of one failure.

A middle ground I'd like to go for is that we import the lines we can, and present a user facing error which says exactly which lines weren't successful and the error that caused it. 

i.e. say I put in a CSV and lines 1,2 and 4 are valid but line 3 gives a PDS error on save (as API call made) and line 5 was missing parts of the CSV we should produce a user message which says sth like 
_"Imported 3 care recipients successfully. 
Error on line 3 (WRIGHT, WANT029193) - Date of birth does not match NHS records - recipient **not imported**
Error on line 5 (YOUNG, WANT038231) - No NHS number provided - recipient **not imported**"_

Reply:
I added some more validation on _rows_ so the broken ones are not imported.

Reply:
Ok, so what's the behaviour now?  I can't see how we're getting the information to the user about which rows failed - we need that `pass` to actually do something.  As it stands the only way the user will know anything's failed at all is if they happen to notice that the reported count doesn't match the number of lines in the file, which is more likely to end up with someone silently getting missed on import.

It would actually be better without the `except` block here.  Can I suggest that we revert back to 
```py
                care_recipient.clean()
                care_recipient.save()
```
without the `try...except` and get another ticket on the board to work this through?